File: app01_java_block/src/main/java/com/gandarych/legalchain/core/BlockCandidate.java
package com.gandarych.legalchain.core;

import java.util.List;

/**
 * An unsealed block handed to a {@link com.gandarych.legalchain.consensus.ConsensusStrategy},
 * which produces the proof/nonce and the final {@link Block}. Separating the candidate
 * from the sealed block is what lets consensus algorithms be swapped (Strategy pattern)
 * without touching the ledger core.
 */
public record BlockCandidate(
        long index,
        long timestamp,
        String previousHash,
        List<Transaction> transactions,
        String validatorId) {

    public String merkleRoot() {
        return Block.merkleRootOf(transactions);
    }

    /** Convenience for strategies: seal this candidate with the given proof and nonce. */
    public Block seal(String consensusType, String proof, long nonce) {
        String merkleRoot = merkleRoot();
        String hash = Block.computeHash(index, timestamp, previousHash, merkleRoot,
                validatorId, consensusType, proof, nonce);
        return new Block(index, timestamp, previousHash, merkleRoot, transactions,
                validatorId, consensusType, proof, nonce, hash);
    }
}

File: app03_csharp_block/src/LegalChain/Core/BlockCandidate.cs
namespace LegalChain.Core;

/// <summary>
/// An unsealed block handed to a <see cref="LegalChain.Consensus.IConsensusStrategy"/>,
/// which produces the proof/nonce and the final <see cref="Block"/>. Separating the
/// candidate from the sealed block is what lets consensus algorithms be swapped
/// (Strategy pattern) without touching the ledger core.
/// </summary>
public record BlockCandidate(
    long Index,
    long Timestamp,
    string PreviousHash,
    IReadOnlyList<Transaction> Transactions,
    string ValidatorId)
{
    public string MerkleRoot() => Block.MerkleRootOf(Transactions);

    /// <summary>Convenience for strategies: seal this candidate with the given proof and nonce.</summary>
    public Block Seal(string consensusType, string proof, long nonce)
    {
        string merkleRoot = MerkleRoot();
        string hash = Block.ComputeHash(Index, Timestamp, PreviousHash, merkleRoot,
            ValidatorId, consensusType, proof, nonce);
        return new Block(Index, Timestamp, PreviousHash, merkleRoot, Transactions,
            ValidatorId, consensusType, proof, nonce, hash);
    }
}
